tests/client/test_client.py

```diff
@@ -165,3 +176,81 @@ def test_get_missing(client, requests_mocker):
 
     # It should explain the problem
     assert "repo1 was not found" in str(error.value)
+
+
+def test_can_get_maintenance_report(client, requests_mocker):
+    maintenance_report = {
+        "last_updated": "2019-08-15T14:21:12Z",
+        "last_updated_by": "Content Delivery",
+        "repos": {
+            "repo1": {
+                "message": "Maintenance Mode Enabled",
+                "owner": "Content Delivery",
+                "started": "2019-08-15T14:21:12Z",
+            }
+        },
+    }
+    requests_mocker.get(
+        "https://pulp.example.com/pulp/isos/redhat-maintenance/repos.json",
+        json=maintenance_report,
+    )
+
+    report = client.get_maintenance_report().result()
+
+    assert requests_mocker.call_count == 1
+    assert isinstance(report, MaintenanceReport)
+    assert report.last_updated_by == "Content Delivery"
+    assert report.last_updated == "2019-08-15T14:21:12Z"
+    assert report.entries[0].message == "Maintenance Mode Enabled"
+
+
+def test_non_maintenance_report(client, requests_mocker):
+    requests_mocker.get(
+        "https://pulp.example.com/pulp/isos/redhat-maintenance/repos.json",
+        text="Not Found",
+        status_code=404,
+    )
+
+    report = client.get_maintenance_report().result()
+    assert report.last_updated_by is None
+    assert report.entries == ()
+
+
+def test_set_maintenance(client, requests_mocker):
+    maintenance_report = {
+        "last_updated": "2019-08-15T14:21:12Z",
+        "last_updated_by": "pubtools.pulplib",
+        "repos": {
+            "repo1": {
+                "message": "Maintenance Mode Enabled",
+                "owner": "pubtools.pulplib",
+                "started": "2019-08-15T14:21:12Z",
+            }
+        },
+    }
+    requests_mocker.post(
+        "https://pulp.example.com/pulp/api/v2/repositories/search/",
+        [{"json": [{"id": "redhat-maintenance", "notes": {"_repo-type": "iso-repo"}}]}],
+    )
+
+    report = MaintenanceReport._from_data(maintenance_report)
+
+    with patch("pubtools.pulplib.FileRepository.upload_file") as mocked_upload:
+        with patch("pubtools.pulplib.Repository.publish") as mocked_publish:
+            mocked_upload.return_value = f_return()
+            mocked_publish.return_value = f_return()
+            client.set_maintenance(report).result()
```

Comment:
Related to my question about what this function actually returns: you never tested the returned value here.  How do you know it doesn't return garbage?

Can you please either test the returned value, or make it return nothing meaningful (e.g. Future[None]).

Reply:
I figured out I was wrong about that other comment.  But the point still seems to be valid that the returned value is never checked here?

Reply:
I'm not saying that it needs a very detailed test.  But right now, you've documented it as returning a Future[list[Task]], and this test only verifies that it returns a Future[?].  So there is no protection at all against it returning something entirely bogus, like Future[None], Future[str] or whatever...

Actually I think it would be fair to simply test it like this:

```suggestion
            f = client.set_maintenance(report)
            f.result()
            # It should return *exactly* what publish returned
            assert f is mocked_publish.return_value
```

As you say, publish is tested elsewhere.  So if we implement this API as "it returns the result of publish", it seems fair to directly test that.

Reply:
Hm, it's not completely the same. 
The set_maintenance returns a FlatMapFuture object, which wraps upload and publish. mocked_publish returns a Future, so they can't be same..., though the results are identical. 

I tried to assert f == f_flat_map(mocked_upload(), lambda _: mocked_publish()), but it gave me this error:
>               assert f == f_flat_map(mocked_upload(), lambda _: mocked_publish())
E               AssertionError: assert <FlatMapFuture at 0x7f7f54c19b50 state=finished returned NoneType> == <FlatMapFuture at 0x7f7f54c19e10 state=finished returned NoneType>
E                +  where <FlatMapFuture at 0x7f7f54c19e10 state=finished returned NoneType> = f_flat_map(<Future at 0x7f7f54c19990 state=finished returned NoneType>, <function <lambda> at 0x7f7f54cadb90>)
E                +    where <Future at 0x7f7f54c19990 state=finished returned NoneType> = <MagicMock name='upload_file' id='140184859088656'>()

Can we check if the result of set_maintenance is whatever wrapped in f_return()?